js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {DataType} from '../../../wasm-common';
+import {TensorView} from '../../tensor-view';
+import {ShapeUtil} from '../../util';
+import {AttributeWithCacheKey, createAttributeWithCacheKey} from '../attribute-with-cache-key';
+import {ComputeContext, ProgramInfo} from '../types';
+
+import {createTensorShapeVariables, IndicesHelper, inputVariable, outputVariable, ShaderHelper} from './common';
+
+export interface FormatAttributes {
+  readonly format: 'NHWC'|'NCHW';
+}
+
+export interface DepthToSpaceAttributes extends FormatAttributes, AttributeWithCacheKey {
+  readonly blocksize: number;
+  readonly mode: string;
+}
+
+const validateInputs = (inputs: readonly TensorView[]): void => {
+  if (!inputs || inputs.length !== 1) {
+    throw new Error('DepthToSpace requires 1 input.');
+  }
+  if (inputs[0].dims.length !== 4) {
+    throw new Error('DepthToSpace requires 4D input.');
+  }
+};
+
+const permFunctionBody = (perm: number[], rank: number, input: IndicesHelper, output: IndicesHelper): string => {
+  const reverseFunc = [];
+  reverseFunc.push(`fn perm(i: ${output.type.indices}) -> ${input.type.indices} {
+    var a: ${input.type.indices};`);
+  for (let i = 0; i < rank; ++i) {
+    reverseFunc.push(input.indicesSet('a', perm[i], `i[${i}]`));
+  }
+  reverseFunc.push('return a;}');
+  return reverseFunc.join('\n');
+};
+
+const createDepthToSpaceProgramInfo = (inputTensor: TensorView, attributes: DepthToSpaceAttributes): ProgramInfo => {
+  let n: number, h: number, w: number, c: number;
+  let shape: number[];
+  let perm: number[];
+  const isChannelLast = attributes.format === 'NHWC';
+  const blocksize = attributes.blocksize;
+  const isDCRmode = attributes.mode === 'DCR';
+  if (isChannelLast) {
+    [n, h, w, c] = inputTensor.dims;
+    shape = isDCRmode ? [n, h, w, blocksize, blocksize, c / (blocksize ** 2)] :
+                        [n, h, w, c / (blocksize ** 2), blocksize, blocksize];
+    perm = isDCRmode ? [0, 1, 3, 2, 4, 5] : [0, 1, 4, 2, 5, 3];
+  } else {
+    [n, h, w, c] = [inputTensor.dims[0], inputTensor.dims[2], inputTensor.dims[3], inputTensor.dims[1]];
+    shape = isDCRmode ? [n, blocksize, blocksize, c / (blocksize ** 2), h, w] :
+                        [n, c / (blocksize ** 2), blocksize, blocksize, h, w];
+    perm = isDCRmode ? [0, 3, 4, 1, 5, 2] : [0, 1, 4, 2, 5, 3];
+  }
+  const reshapedInputTensor = inputTensor.reshape(shape);
+
+  const shapeBeforePerm = reshapedInputTensor.dims;
+  const shapeAfterPerm = ShapeUtil.sortBasedOnPerm(shapeBeforePerm, perm);
+
+  const inputDataType = inputTensor.dataType;
+  const reshapedInput = inputVariable('a', inputDataType, shapeBeforePerm);
+  const permedOutput = outputVariable('output', inputDataType, shapeAfterPerm);
```

Comment:
There is a mismatch between the usage of `inputVariable`/`outputVariable` functions and the uniform data.

There are 2 ways using the 3rd parameter of `inputVariable`/`outputVariable` functions:

- using a shape (number[]), so the shape info will be hardcoded in the source
- using a rank (number), so the shape info is expected to be included in the `programUniforms` property returned by `getRunData()` function.

I assume that you want to use uniform, so changing the shape to rank here can fix this problem.

Reply:
I've understood the programUniforms functionality thanks to your explanations.
I changed the shape to rank in the additional commit.